### PR TITLE
fix(client): forbid single-part domains in email addresses

### LINKED ISSUE #2199: Reject email addresseses without a dot in the domain component
When testing a form I entered in an intentionally invalid email syntax, but was able to proceed. The form should not submit until a valid email is entered (and messages should appear)
whoneeds

### PR DESCRIPTION
**Pull request by philbooth**
Fixes #2199.

~~I tried upgrading to the latest version of Gerv's regex but I couldn't make it work. Either regexes are hard or I'm stupid, or possibly both.~~

~~So instead, this change just modifies our existing regex to enforce at least one dot on the domain side of the `@`.~~ I also took the liberty of simplifying the length checks by moving them into the regex because the call to `split` seemed unnecessary to me. And I flipped the order of the tests round so that we perform the cheapest one first.

There's a few new tests added to assert that the length-based stuff didn't break and of course some other tests had to change now that we're rejecting emails we were previously accepting.

@shane-tomlinson, @vladikoff, @vbudhram, can I get a ~~witness~~ r?

### CODE REVIEW

File: app/tests/spec/views/sign_up.js [hunk 1]
```diff
@@ -528,10 +528,6 @@ define(function (require, exports, module) {
         view.showValidationErrors();
 
         assert.isTrue(view.showValidationError.called);
-
-        var err = AuthErrors.toError('DIFFERENT_EMAIL_REQUIRED_FIREFOX_DOMAIN');
```
Comment (vbudhram) [inline app/tests/spec/views/sign_up.js]: Does this mean that it doesn't log the error if using @firebox domain?

File: app/tests/spec/lib/validate.js [hunk 2]
```diff
@@ -16,16 +16,36 @@ define(function (require, exports, module) {
 
   describe('lib/validate', function () {
     describe('isEmailValid', function () {
+      it('returns true for valid email', function () {
```
Comment (vbudhram) [inline app/tests/spec/lib/validate.js]: Not sure if we want to cover this edge case but joi validates emails like `joe@[IPv6:2a00:1450:4001:c02::1b]`. The current regex fails for this.

https://github.com/hapijs/joi/blob/master/test/string.js#L471

**Comment by jrgm** [comment]
Maybe we should use something like https://github.com/mozilla/fxa-auth-server/blob/master/lib/routes/validators.js#L40-L93 (derived from what the browser itself does) on the front-end code and make it so that the back-end and front-end share the same implementation. 

One dependency that needs to be satisfied is for `punycode`, but the code in node.js' `punycode` is based on work by Mathias Bynens, who also a generic implementation that should be usable in the front-end (maybe via bower, or we can just vendor the impl).

We're probably overdue on doing a compare and possible refresh of the above code to pick up any relevant code changes from the gecko/firefox version. 

Just a thought. Could be a very stupid thought even.

**Comment by philbooth** [comment]
That sounds like a pretty good suggestion to me, thanks @jrgm. I'll double-check with @shane-tomlinson tomorrow before implementing.

File: app/tests/spec/views/sign_up.js (hunk 1 above)
Comment (philbooth) [inline app/tests/spec/views/sign_up.js]: Yep. It was purely out of simplicity though, if it's a problem I'm happy to do something about it.

**Comment by philbooth** [comment]
So, after some discussion with @shane-tomlinson in IRC, I'm sticking with the regex-based approach in order to avoid increasing the front-end payload size.

However, @jrgm's point about consistency of implementation is a good one, so I've imported unit tests from the auth server to ensure that we're at least demonstrating parity with those test cases. An added bonus from passing those tests was that it necessitated using the latest version of Gerv's regex, which is now working as we want it to (I should have tried harder at that on Friday, sorry).

And as a further step towards consistency of implementation, I'm going to try PRing https://github.com/mozilla/fxa-auth-server/issues/1193 to use the regex from here. Not sure how the back-end folks will feel about that but at least we know we pass their current test suite.

@shane-tomlinson, @vbudhram, care to take another look at this?

File: app/tests/spec/lib/validate.js [hunk 3]
```diff
@@ -16,16 +16,62 @@ define(function (require, exports, module) {
...
+          'notAnEmailAddress',
+          '\n@example.com',
+          'me@hello world.com',
+          'me@hello+world.com',
+          'me@.example',
+          'me@example.com-',
+          'me@example..com',
+          'me@example-.com',
+          'me@example.-com'
+        ].forEach(function (email) {
```
Comment (philbooth) [inline app/tests/spec/lib/validate.js]: If we're really serious about maintaining parity of behaviour between the content server and the auth server, these test cases (or the implementation itself, if we get to the point of re-using it as part of https://github.com/mozilla/fxa-auth-server/issues/1193) could be moved out to a dedicated repo and submoduled in.

Would that be overkill for something so small? The duplication feels wrong is all.

**Comment by philbooth** [comment]
Sorry, ignore this. Just noticed there are some auth server tests that I didn't import and they aren't passing. Closing until I have a passing implementation.

**Comment by rfk** [comment]
> make it so that the back-end and front-end share the same implementation. 

See also https://github.com/mozilla/fxa-auth-server/issues/1015 were we talk about extracting that logic into a re-usable implementation.
